File: integracoes_customizadas/juridico/report/relação_de_processos/relação_de_processos.py
"""Relação consultável dos processos judiciais cadastrados."""

from collections import defaultdict

import frappe
from frappe import _
from frappe.utils import getdate


def execute(filters=None):
	filters = frappe._dict(filters or {})
	query_filters = []
	if filters.get("company"):
		query_filters.append(["Processo Judicial", "empresa", "=", filters.company])
	for field in (
		"tipo_parte", "parte", "fase_processo", "status_processo",
		"situacao_cadastro", "risco", "tribunal", "grau",
	):
		if filters.get(field):
			query_filters.append(["Processo Judicial", field, "=", filters[field]])

	if filters.get("numero_processo"):
		query_filters.append(["Processo Judicial", "numero_processo", "like", f"%{filters.numero_processo}%"])
	if filters.get("assunto"):
		query_filters.append(["Processo Judicial Assunto", "assunto", "=", filters.assunto])
	if filters.get("de"):
		query_filters.append(["Processo Judicial", "data_ajuizamento", ">=", getdate(filters.de)])
	if filters.get("ate"):
		query_filters.append(["Processo Judicial", "data_ajuizamento", "<", frappe.utils.add_days(getdate(filters.ate), 1)])
	if filters.get("de") and filters.get("ate") and getdate(filters.de) > getdate(filters.ate):
		frappe.throw(_("A data inicial não pode ser posterior à data final."))

	fields = [
		"name", "numero_processo", "empresa", "tipo_parte", "parte", "nome_parte",
		"fase_processo", "status_processo", "situacao_cadastro", "risco", "tribunal",
		"grau", "orgao_julgador_nome", "classe_nome", "data_ajuizamento",
		"data_hora_ultima_atualizacao", "valor_causa", "valor_condenacao", "moeda",
	]
	# get_list aplica as permissões do DocType, inclusive ao filtrar pela tabela filha.
	data = frappe.get_list(
		"Processo Judicial", fields=fields, filters=query_filters,
		order_by="data_ajuizamento desc, name asc", limit_page_length=0, distinct=True,
	)
	assuntos = defaultdict(list)
	for start in range(0, len(data), 500):
		names = [row.name for row in data[start:start + 500]]
		for item in frappe.get_all(
			"Processo Judicial Assunto",
			filters={"parent": ["in", names], "parenttype": "Processo Judicial"},
			fields=["parent", "nome", "assunto"], order_by="idx asc",
		):
			assuntos[item.parent].append(item.nome or item.assunto)
	for row in data:
		row.nome_parte = row.nome_parte or row.parte
		row.assuntos = ", ".join(assuntos[row.name])
	return get_columns(), data


def get_columns():
	return [
		{"fieldname": "numero_processo", "label": _("Número do Processo"), "fieldtype": "Data", "width": 195},
		{"fieldname": "empresa", "label": _("Empresa"), "fieldtype": "Link", "options": "Company", "width": 180},
		{"fieldname": "nome_parte", "label": _("Parte"), "fieldtype": "Data", "width": 190},
		{"fieldname": "fase_processo", "label": _("Fase"), "fieldtype": "Data", "width": 110},
		{"fieldname": "status_processo", "label": _("Situação"), "fieldtype": "Data", "width": 125},
		{"fieldname": "tribunal", "label": _("Tribunal"), "fieldtype": "Data", "width": 95},
		{"fieldname": "classe_nome", "label": _("Classe"), "fieldtype": "Data", "width": 200},
		{"fieldname": "assuntos", "label": _("Assuntos"), "fieldtype": "Data", "width": 230},
		{"fieldname": "data_ajuizamento", "label": _("Ajuizamento"), "fieldtype": "Datetime", "width": 150},
		{"fieldname": "risco", "label": _("Risco"), "fieldtype": "Data", "width": 90},
		{"fieldname": "valor_causa", "label": _("Valor da Causa"), "fieldtype": "Currency", "options": "moeda", "width": 140},
	]

File: integracoes_customizadas/juridico/report/relação_de_processos/relação_de_processos.js
frappe.query_reports["Relação de Processos"] = {
	filters: [
		{ fieldname: "company", label: __("Empresa"), fieldtype: "Link", options: "Company" },
		{
			fieldname: "tipo_parte", label: __("Tipo da Parte"), fieldtype: "Select",
			options: "\nEmployee\nCustomer\nSupplier",
			on_change: () => frappe.query_report.set_filter_value("parte", ""),
		},
		{
			fieldname: "parte", label: __("Parte"), fieldtype: "Dynamic Link",
			options: "tipo_parte", depends_on: "eval:doc.tipo_parte",
		},
		{
			fieldname: "fase_processo", label: __("Fase"), fieldtype: "Select",
			options: "\nPostulatória\nInstrutória\nDecisória\nRecursal\nExecução",
		},
		{
			fieldname: "status_processo", label: __("Situação"), fieldtype: "Select",
			options: "\nEm andamento\nEncerrado",
		},
		{ fieldname: "assunto", label: __("Assunto"), fieldtype: "Link", options: "Assunto Judicial" },
		{ fieldname: "tribunal", label: __("Tribunal"), fieldtype: "Data" },
		{ fieldname: "numero_processo", label: __("Número do Processo"), fieldtype: "Data" },
		{
			fieldname: "situacao_cadastro", label: __("Situação do Cadastro"), fieldtype: "Select",
			options: "\nRascunho\nCadastrado",
		},
		{ fieldname: "risco", label: __("Risco"), fieldtype: "Select", options: "\nBaixo\nMédio\nAlto" },
		{ fieldname: "grau", label: __("Grau"), fieldtype: "Data" },
		{ fieldname: "de", label: __("Ajuizado de"), fieldtype: "Date" },
		{ fieldname: "ate", label: __("Ajuizado até"), fieldtype: "Date" },
	],
	html_format: `
		<style>
			.relacao-processos .processo { break-inside: avoid; page-break-inside: avoid; margin: 0 0 15px; }
			.relacao-processos .processo h3 { font-size: 13px; margin: 0 0 6px; padding: 5px 8px; background: #f1f3f5; }
			.relacao-processos .processo table { width: 100%; border-collapse: collapse; font-size: 10px; }
			.relacao-processos .processo td { padding: 4px 7px; border: 1px solid #d9dee3; vertical-align: top; width: 50%; }
			.relacao-processos .rotulo { color: #586069; display: block; font-size: 9px; }
		</style>
		<div class="relacao-processos">
			<h2 class="text-center">{{ __(title) }}</h2>
			<p class="text-center">{{ data.length }} {{ __("processo(s)") }}</p>
			{% if subtitle %}<div>{{ subtitle }}</div>{% endif %}
			{% for row in data %}
			<div class="processo">
				<h3>{{ frappe.utils.escape_html(row.numero_processo || row.name) }}</h3>
				<table>
					<tr><td><span class="rotulo">{{ __("Empresa") }}</span>{{ frappe.utils.escape_html(row.empresa || "—") }}</td><td><span class="rotulo">{{ __("Parte") }}</span>{{ frappe.utils.escape_html(row.nome_parte || row.parte || "—") }}</td></tr>
					<tr><td><span class="rotulo">{{ __("Fase / Situação") }}</span>{{ frappe.utils.escape_html([row.fase_processo, row.status_processo].filter(Boolean).join(" / ") || "—") }}</td><td><span class="rotulo">{{ __("Tribunal / Grau") }}</span>{{ frappe.utils.escape_html([row.tribunal, row.grau].filter(Boolean).join(" / ") || "—") }}</td></tr>
					<tr><td><span class="rotulo">{{ __("Classe") }}</span>{{ frappe.utils.escape_html(row.classe_nome || "—") }}</td><td><span class="rotulo">{{ __("Órgão julgador") }}</span>{{ frappe.utils.escape_html(row.orgao_julgador_nome || "—") }}</td></tr>
					<tr><td colspan="2"><span class="rotulo">{{ __("Assuntos") }}</span>{{ frappe.utils.escape_html(row.assuntos || "—") }}</td></tr>
					<tr><td><span class="rotulo">{{ __("Ajuizamento") }}</span>{{ row.data_ajuizamento ? frappe.format(row.data_ajuizamento, {fieldtype: "Datetime"}) : "—" }}</td><td><span class="rotulo">{{ __("Risco / Valor da Causa") }}</span>{{ frappe.utils.escape_html(row.risco || "—") }} / {{ row.valor_causa != null ? frappe.format(row.valor_causa, {fieldtype: "Currency", options: "moeda"}, {moeda: row.moeda}) : "—" }}</td></tr>
					<tr><td><span class="rotulo">{{ __("Situação do Cadastro / Tipo da Parte") }}</span>{{ frappe.utils.escape_html([row.situacao_cadastro, row.tipo_parte].filter(Boolean).join(" / ") || "—") }}</td><td><span class="rotulo">{{ __("Valor da Condenação") }}</span>{{ row.valor_condenacao != null ? frappe.format(row.valor_condenacao, {fieldtype: "Currency", options: "moeda"}, {moeda: row.moeda}) : "—" }}</td></tr>
				</table>
			</div>
			{% endfor %}
		</div>
	`,
};
